File: mpi/docs/mpi_character/example_2.f90
!> @brief Illustrate how to communicate a string between 2 MPI processes.
!> @details This application is meant to be run with 2 MPI processes: 1 sender
!> and 1 receiver. The former sends a string to the latter, which prints it.
PROGRAM main
    USE mpi

    IMPLICIT NONE

    INTEGER :: ierror
    INTEGER :: size
    INTEGER, PARAMETER :: sender_rank = 0
    INTEGER, PARAMETER :: receiver_rank = 1
    INTEGER :: my_rank
    CHARACTER(LEN=11) :: stringToSend
    CHARACTER(LEN=11) :: stringReceived

    CALL MPI_Init(ierror)

    ! Check that 2 MPI processes are used.
    CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierror)
    IF (size .NE. 2) THEN
        WRITE(*,'(A)') 'This application is meant to be run with 2 MPI processes.'
        CALL MPI_Abort(MPI_COMM_WORLD, -1, ierror)
    END IF

    ! Get my rank and do the corresponding job.
    CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank, ierror)
    SELECT CASE (my_rank)
        CASE (sender_rank)
            ! Sends the string
            stringToSend = 'Hello world'
            WRITE(*,'(A,I0,A,A,A)') '[MPI process ', my_rank, '] I send string: "', stringToSend, '".'
            CALL MPI_Ssend(stringToSend, 11, MPI_CHARACTER, receiver_rank, 0, MPI_COMM_WORLD, ierror)
        CASE (receiver_rank)
            ! Receives the string
            CALL MPI_Recv(stringReceived, 11, MPI_CHARACTER, sender_rank, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror)
            WRITE(*,'(A,I0,A,A,A)') '[MPI process ', my_rank, '] I received string: "', stringReceived, '".'
    END SELECT

    CALL MPI_Finalize(ierror)
END PROGRAM main

File: mpi/docs/mpi_long/example_2.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

/**
 * @brief Illustrate how to communicate an array of long ints between 2 MPI
 * processes.
 * @details This application is meant to be run with 2 MPI processes: 1 sender
 * and 1 receiver. The former sends an array of long ints to the latter, which
 * prints it.
 **/
int main(int argc, char* argv[])
{
	MPI_Init(&argc, &argv);

	// Check that 2 MPI processes are used.
	int size;
	MPI_Comm_size(MPI_COMM_WORLD, &size);
	if(size != 2)
	{
		printf("This application is meant to be run with 2 MPI processes.\n");
		MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
	}

	// Get my rank and do the corresponding job.
	enum role_ranks { SENDER, RECEIVER };
	int my_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
	switch(my_rank)
	{
		case SENDER:
		{
			// Send the long ints
			long int longIntsToSend[2] = { 12345, 67890 };
			printf("[MPI process %d] I send long ints: %ld and %ld.\n", my_rank, longIntsToSend[0], longIntsToSend[1]);
			MPI_Ssend(longIntsToSend, 2, MPI_LONG, RECEIVER, 0, MPI_COMM_WORLD);
			break;
		}
		case RECEIVER:
		{
			// Receive the long ints
			long int longIntsReceived[2];
			MPI_Recv(longIntsReceived, 2, MPI_LONG, SENDER, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
			printf("[MPI process %d] I received long ints: %ld and %ld.\n", my_rank, longIntsReceived[0], longIntsReceived[1]);
			break;
		}
	}

	MPI_Finalize();

	return EXIT_SUCCESS;
}
